enhancements/hypershift/topology-aware-scheduling/topology-aware-scheduling.md

```diff
@@ -0,0 +1,256 @@
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+This proposal aims for Hypershift / Hosted Control Planes alone. 
+
+#### Standalone Clusters
+
+* N/A
+
+#### Single-node Deployments or MicroShift
+
+* N/A
+
+### Implementation Details/Notes/Constraints
+
+#### Expose KubeletConfig to Hosted-Clusters
+
+At first, when a user associates a KubeletConfig's ConfigMap to the NodePool,
+the NodePool controller mirrors the ConfigMap to the control-plane namespace.
+Similar to: https://github.com/openshift/hypershift/pull/4150
+The KubeletConfig's ConfigMap NamespacedName convention, created by the NodePool controller is:
+`<control-plane-namespace>/<name+node-pool>`
+* `control-plane-namespace` = name of the control-plane namespace
+* `name` = name of the ConfigMap as provided by the user in the hosted-cluster namespace.
+* `node-pool` = name of the NodePool on which the ConfigMap was referenced.
+
+To make sure the final name is valid,
+it's given to the https://github.com/openshift/hypershift/blob/ffa435f71e1a037174e29d31a2b80d569c940d57/support/util/route.go#L39 
+function so that 
+```
+finalName := ShortenName(name, NodePoolName)
+```
+In addition, the NodePool controller creates the KubeletConfig ConfigMap with the following labels:
+* `hypershift.openshift.io/kubeletconfig-config` : `""`
+* `hypershift.openshift.io/nodePool` : `NodePool` API name where the KubeletConfig has referenced
+
+Then, HCCO picks it up from the control-plane namespace, 
+by listing ConfigMaps with the labels above, and copies it to the hosted-cluster.
+
+The KubeletConfig's ConfigMap NamespacedName convention, created by HCCO is:
+`openshift-config-managed/<name>`
+* `openshift-config-managed` = name of the namespace where all the configs managed by OCP reside.
+* `name` = name of the ConfigMap as provided by the NodePool controller in the control-plane namespace.
+
+When the KubeletConfig ConfigMap is deleted by the user (removed from the NodePool spec), 
+the NodePool controller deletes the ConfigMap from the control-plane namespace.
+HCCO will watch for changes in ConfigMaps under the CP NS:
+```
+p := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		cm := o.(*corev1.ConfigMap)
+		if _, ok := cm.Labels[nodepool.KubeletConfigConfigMapLabel]; ok {
+			return true
+		}
+		return false
+	})
+	if err := c.Watch(source.Kind[client.Object](opts.CPCluster.GetCache(), &corev1.ConfigMap{}, eventHandler(), p)); err != nil {
+		return fmt.Errorf("failed to watch ConfigMap: %w", err)
+	}
+```
+so the deletion of the ConfigMaps triggers the reconciliation loop of HCCO.
+HCCO compares the KubeletConfig ConfigMaps on the CP NS against the KubeletConfig ConfigMaps on the
+hosted cluster, and deletes the ones exist on the hosted cluster but are no longer on the CP NS. 
+
+It is possible to have multiple KubeletConfig's ConfigMap (one per each NodePool) under the same hosted cluster.
```

Comment:
what happens if there's references to multiple kubeletConfigs?

Reply:
It should not be allowed. 
It's also not allowed on standard OCP, MCO ignores the second KubeletConfig.
I'll state that in the enhancement. 

We have a choice here, in case user is referring to more than a single KubeletConifg, HCCO can:
* log a warning, ignore the second KubeletConfig and keep operate as usual
OR
* Throw an error, and requires user intervention to remove the additional KubeletConifg in order to make the operator functional again.

I personally prefer the second option since it's more restrict, and we already have similar logic with other objects (see: https://github.com/openshift/hypershift/blob/5ac54b9bb7faec05e780a128c5f97e19fd44a678/hypershift-operator/controllers/nodepool/nodepool_controller.go#L3568) , but i'll leave the decision to you.

Reply:
Yes, hard fail on validation for anything that prevents successful operational behaviour makes sense to me. Let's elaborate on that and anything else that involves needed validation